aeron-driver/src/main/c/aeron_windows.c

```diff
@@ -31,62 +31,6 @@
 #define __builtin_popcount __popcnt
 #define __builtin_popcountll __popcnt64
 
-static DWORD dwTlsIndex; // address of shared memory
-
-// DllMain() is the entry-point function for this DLL.
-
-BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
```

Comment:
Why remove DllMain?

Reply:
It does nothing useful, and wastes a TLS slot.

This code is basically taken from [the docs example](https://docs.microsoft.com/en-us/windows/win32/dlls/using-thread-local-storage-in-a-dynamic-link-library) but unlike the example it does not use `dwTlsIndex` anywhere outside of `DllMain`. So basically `DllMain` manages a TLS slot and some per-thread memory just for itself. The other TLS code does not need this.

Reply:
Actually, I think DllMain will be necessary. Currently, the error handling is broken in tests due to the TLS state not working. And that is partially due to it not being wired in correctly with DllMain or handling static libs correctly.

Reply:
A correct DllMain should use the existing slot and make the checks work. The changes to the TLS handling may work for static libs. Should write a test to make sure error handling is actually correct for threading, etc.

Reply:
I actually fixed the error handling in this PR, it was previously causing one of the unit tests to fail because `aeron_thread_once` was not working properly.

While we *could* make the TLS slot be managed by `DllMain`, it would break TLS if the media driver is used as a static library, since those don't use `DllMain`, and it would create the need to maintain different code paths for Windows and Linux.

I think the TLS code from this PR is fine, and it doesn't need a `DllMain` function to work properly, unless I'm misunderstanding something.

Reply:
You don't need different code paths. The DllMain can just pre-populate the TLS key/slot. Same code in what is there now but with DllMain but thread once will need to see if the slot is pre-populate or not. And as you point out it is needed for static libs. 

A DllMain is needed as we have plans that do require it soon. Also, while it is small, it is an advantage to pre-populate on DLL attach when possible.

Reply:
> And as you point out it is needed for static libs.

Not exactly, I just said that `DllMain` is not called in static libs, which would require a different initialization path if you want to support them.

But since you have plans to require `DllMain` soon, I suppose that using the media driver as a static library will not be possible anymore in the future, right?

I could move the TLS slot handling to `DllMain` but my understanding is that it will make retrieving the error messages impossible in a static library on Windows as a consequence, unless we add the currently existing code in an `#if STATIC_LIBRARY` or something similar.

Reply:
This can be done, I think, quite easily. A static set to indicate that `TlsAlloc` has been called by DllMain means that `aeron_thread_key_create` doesn't need to call it. If not, then it can call `TlsAlloc`. All of this can be hidden in there. So, this shouldn't require anything weird. The use of `TlsSetValue` can still use the existing paths and created on demand.

If you are not comfortable about this or do not quite understand what I am suggesting. Just keep the DllMain and remove the TLS calls from it and keep the rest as you fixed it. I'll fix DllMain up when I have time.

In terms of requiring DllMain. Only the use of specific options may (or may not). Eventually, we would like to do DPDK, RIO, and other specific support. Some of that will be easier to integrate with DllMain than without.

Reply:
BTW, in a static lib, won't the error state be a leak when the thread goes away? Meaning to avoid the leak, DLL would be a better idea? (I've not looked for a hook to catch thread death for a while)...

Reply:
Ok I think I understand what you're suggesting now. I'll implement that tomorrow.

> Eventually, we would like to do DPDK, RIO, and other specific support.

That's great! 👍

> Some of that will be easier to integrate with DllMain than without.

I figured an empty `DllMain` wouldn't be useful and you could add it back at a later time when needed. Anyway, I'll add it back with specific support for error messages.

> BTW, in a static lib, won't the error state be a leak when the thread goes away?

Yes, it will leak. That can be mitigated from `DllMain`. But if I understand correctly, it also means there's currently a leak on Linux.

Reply:
On Linux it doesn't leak because `pthread_key_create` takes `free` as the destructor arg.